### DISCUSSION TITLE
Build -> Core -> Nodebb(Problem within the docker container used in build) - Build Failed

### DISCUSSION THREAD

**Discussion by shounakacharya** [Sunbird-Ed/Community#80] [Installation] [answered]
System Configuration Details:
1.	KP, DP, DB, Yarn, Jenkins, Other-Services: Ubuntu 18.04 LTS, 4vCPU, 16GB RAM, 250 GB HDD
2.	Google Kubernetes engine: 1.21.12-gke.2200 with 4 worker nodes. Note: Currently Sunbird is not compatible with K8s APIs  >= 1.22 and we have checked this
3.	All systems are running on GCP
4. Github Tag: github_release_tag: release-4.6.0_RC3, nodebb_branch: v1.16.0

@gandham-santhosh and  @keshavprasadms

Jenkins Logs:

Started by user [Admin Deployer](http://35.222.148.131:8080/user/admin)
Checking out git https://github.com/project-sunbird/sunbird-nodebb.git into /var/lib/jenkins/workspace/Build/Core/Nodebb@script to read Jenkinsfile
The recommended git tool is: NONE
No credentials specified
Cloning the remote Git repository
Cloning repository https://github.com/project-sunbird/sunbird-nodebb.git
 > git init /var/lib/jenkins/workspace/Build/Core/Nodebb@script # timeout=10
Fetching upstream changes from https://github.com/project-sunbird/sunbird-nodebb.git
 > git --version # timeout=10
 > git --version # 'git version 2.17.1'
 > git fetch --tags --progress -- https://github.com/project-sunbird/sunbird-nodebb.git +refs/heads/*:refs/remotes/origin/* # timeout=10
 > git config remote.origin.url https://github.com/project-sunbird/sunbird-nodebb.git # timeout=10
 > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
Avoid second fetch
 > git rev-parse origin/release-4.6.0_RC3^{commit} # timeout=10
 > git rev-parse release-4.6.0_RC3^{commit} # timeout=10
Checking out Revision 491c1d52fc9d6aceb9da87fc28325267f7910fe9 (release-4.6.0_RC3)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 491c1d52fc9d6aceb9da87fc28325267f7910fe9 # timeout=10
Commit message: "Issue #SB-28051 merge: Merge pull request #31 from VenkatKadiveti/r4.6-defect"
First time build. Skipping changelog.
Running in Durability level: MAX_SURVIVABILITY
[Pipeline] Start of Pipeline
[Pipeline] node
Running on [Jenkins](http://35.222.148.131:8080/computer/(built-in)/) in /var/lib/jenkins/workspace/Build/Core/Nodebb
[Pipeline] {
[Pipeline] ansiColor
[Pipeline] {

[Pipeline] stage
[Pipeline] { (Checkout)
[Pipeline] echo
Found environment variable named hub_org with value as: shounakacharya/sunbird-setup
[Pipeline] cleanWs
[WS-CLEANUP] Deleting project workspace...
[WS-CLEANUP] Deferred wipeout is used...
[WS-CLEANUP] done
[Pipeline] checkout
The recommended git tool is: NONE
No credentials specified
Cloning the remote Git repository
Cloning repository https://github.com/project-sunbird/sunbird-nodebb.git
 > git init /var/lib/jenkins/workspace/Build/Core/Nodebb # timeout=10
Fetching upstream changes from https://github.com/project-sunbird/sunbird-nodebb.git
 > git --version # timeout=10
 > git --version # 'git version 2.17.1'
 > git fetch --tags --progress -- https://github.com/project-sunbird/sunbird-nodebb.git +refs/heads/*:refs/remotes/origin/* # timeout=10
 > git config remote.origin.url https://github.com/project-sunbird/sunbird-nodebb.git # timeout=10
 > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
Avoid second fetch
 > git rev-parse origin/release-4.6.0_RC3^{commit} # timeout=10
 > git rev-parse release-4.6.0_RC3^{commit} # timeout=10
Checking out Revision 491c1d52fc9d6aceb9da87fc28325267f7910fe9 (release-4.6.0_RC3)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 491c1d52fc9d6aceb9da87fc28325267f7910fe9 # timeout=10
Commit message: "Issue #SB-28051 merge: Merge pull request #31 from VenkatKadiveti/r4.6-defect"
[Pipeline] checkout
The recommended git tool is: NONE
No credentials specified
 > git rev-parse --resolve-git-dir /var/lib/jenkins/workspace/Build/Core/Nodebb/.git # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url https://github.com/project-sunbird/sunbird-nodebb.git # timeout=10
Fetching upstream changes from https://github.com/project-sunbird/sunbird-nodebb.git
 > git --version # timeout=10
 > git --version # 'git version 2.17.1'
 > git fetch --tags --progress -- https://github.com/project-sunbird/sunbird-nodebb.git +refs/heads/*:refs/remotes/origin/* # timeout=10
 > git rev-parse origin/release-4.6.0_RC3^{commit} # timeout=10
 > git rev-parse release-4.6.0_RC3^{commit} # timeout=10
Checking out Revision 491c1d52fc9d6aceb9da87fc28325267f7910fe9 (release-4.6.0_RC3)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 491c1d52fc9d6aceb9da87fc28325267f7910fe9 # timeout=10
Commit message: "Issue #SB-28051 merge: Merge pull request #31 from VenkatKadiveti/r4.6-defect"
[Pipeline] sh
+ echo release-4.6.0_RC3_1
[Pipeline] echo
github_release_tag specified, building from github_release_tag: release-4.6.0_RC3
[Pipeline] echo
build_tag: release-4.6.0_RC3_1
[Pipeline] }
[Pipeline] // stage
[Pipeline] stage
[Pipeline] { (Build)
[Pipeline] sh
+ git clone https://github.com/NodeBB/NodeBB.git -b v1.16.0
Cloning into 'NodeBB'...
Note: checking out '1bece2b79ee4a5ed377127167c07c777f6f02ab9'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

[Pipeline] sh
+ cp Dockerfile NodeBB
[Pipeline] sh
+ cp build.sh NodeBB
[Pipeline] echo
Environment will be : build
[Pipeline] sh
+ chmod 777 NodeBB/build.sh
[Pipeline] sh
+ bash ./NodeBB/build.sh release-4.6.0_RC3_1_v1.16.0 master shounakacharya/sunbird-setup
Sending build context to Docker daemon  102.3MB

Step 1/17 : FROM node:lts
lts: Pulling from library/node
7e6a53d1988f: Pulling fs layer
4fe4e1c58b4a: Pulling fs layer
cc915d298757: Pulling fs layer
f08b88f29371: Pulling fs layer
28bcb226aca8: Pulling fs layer
4ed42ef80d39: Pulling fs layer
bebe51b791bc: Pulling fs layer
986a235ce43f: Pulling fs layer
92cdf774b883: Pulling fs layer
28bcb226aca8: Waiting
4ed42ef80d39: Waiting
bebe51b791bc: Waiting
92cdf774b883: Waiting
986a235ce43f: Waiting
f08b88f29371: Waiting
4fe4e1c58b4a: Verifying Checksum
4fe4e1c58b4a: Download complete
cc915d298757: Verifying Checksum
cc915d298757: Download complete
7e6a53d1988f: Verifying Checksum
7e6a53d1988f: Download complete
4ed42ef80d39: Verifying Checksum
4ed42ef80d39: Download complete
f08b88f29371: Verifying Checksum
f08b88f29371: Download complete
986a235ce43f: Verifying Checksum
986a235ce43f: Download complete
92cdf774b883: Verifying Checksum
92cdf774b883: Download complete
bebe51b791bc: Verifying Checksum
bebe51b791bc: Download complete
28bcb226aca8: Verifying Checksum
28bcb226aca8: Download complete
7e6a53d1988f: Pull complete
4fe4e1c58b4a: Pull complete
cc915d298757: Pull complete
f08b88f29371: Pull complete
28bcb226aca8: Pull complete
4ed42ef80d39: Pull complete
bebe51b791bc: Pull complete
986a235ce43f: Pull complete
92cdf774b883: Pull complete
Digest: sha256:1ed1e17ccabb09038cfb8a965337ebcda51ef9e9d32082164c502d44d9731a02
Status: Downloaded newer image for node:lts
 ---> 954cea825b78
Step 2/17 : RUN mkdir -p /usr/src/app
 ---> Running in e9fdb0cfd6f7
Removing intermediate container e9fdb0cfd6f7
 ---> 7095bcdb5873
Step 3/17 : WORKDIR /usr/src/app
 ---> Running in 3eaecbf970dd
Removing intermediate container 3eaecbf970dd
 ---> 63a3c38c2ad2
Step 4/17 : ARG NODE_ENV
 ---> Running in 9386b49e53b0
Removing intermediate container 9386b49e53b0
 ---> 522ad1ff48a7
Step 5/17 : ENV NODE_ENV $NODE_ENV
 ---> Running in beeb29829b2c
Removing intermediate container beeb29829b2c
 ---> fee7a6b9fa63
Step 6/17 : COPY NodeBB/install/package.json /usr/src/app/package.json
 ---> e9bd4cfe4b27
Step 7/17 : COPY NodeBB/ /usr/src/app
 ---> f19cef2d8f58
Step 8/17 : RUN npm install --only=prod &&     npm cache clean --force
 ---> Running in 378716b39796
npm WARN config only Use `--omit=dev` to omit dev dependencies from the install.
npm notice 
npm notice New minor version of npm available! 8.11.0 -> 8.17.0
npm notice Changelog: <https://github.com/npm/cli/releases/tag/v8.17.0>
npm notice Run `npm install -g npm@8.17.0` to update!
npm notice 
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: nodebb@1.16.0
npm ERR! Found: nodebb-plugin-emoji@3.5.17
npm ERR! node_modules/nodebb-plugin-emoji
npm ERR!   nodebb-plugin-emoji@"^3.3.0" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer nodebb-plugin-emoji@"^2.0.0" from nodebb-plugin-emoji-android@2.0.0
npm ERR! node_modules/nodebb-plugin-emoji-android
npm ERR!   nodebb-plugin-emoji-android@"2.0.0" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /root/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2022-08-16T16_31_18_862Z-debug-0.log
The command '/bin/sh -c npm install --only=prod &&     npm cache clean --force' returned a non-zero code: 1
[Pipeline] }
[Pipeline] // stage
[Pipeline] }

[Pipeline] // ansiColor
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
ERROR: script returned exit code 1
Finished: FAILURE

**Comment by beepdot** [comment, marked as answer]
@vinukumar-vs Can you please provide a solution for this? Has this issue been resolved in new version tags? Similar issues were raised earlier
https://github.com/project-sunbird/sunbird-community/discussions/645
https://github.com/project-sunbird/sunbird-community/discussions/676

**Reply by shounakacharya** [reply]
@keshavprasadms and @vinukumar-vs, any updates on this?

**Reply by beepdot** [reply]
I don't think there will be a fix in this older release version due to time constraints. Please use below image and retag as per your docker hub repo and push it to your repo.
```
docker pull keshavprasad/nodebb:release-4.6.0_RC3_1_v1.18.6
```
Please run the job and allow it to fail and then comment this line in jenkins workspace - https://github.com/Sunbird-Lern/sunbird-nodebb/blob/release-4.6.0/build.sh#L11
In the last failed job on Jenkins, please click replay and comment out the following lines
- https://github.com/Sunbird-Lern/sunbird-nodebb/blob/release-4.6.0/Jenkinsfile#L17
- https://github.com/Sunbird-Lern/sunbird-nodebb/blob/release-4.6.0/Jenkinsfile#L19
After this the ArtifactUpload job should get triggered.

**Reply by beepdot** [reply]
While running the Build job, please use `v1.18.6` as the value for `nodebb_branch` in Jenkins job parameter and `github_release_tag` as `release-4.6.0_RC3`.

**Reply by shounakacharya** [reply]
@keshavprasadms : I followed your steps, but the build job succeeded and uploaded the nodebb docker image to repository in the artifactupload step. I did not have to manually replay the job. However, when I deployed nodebb, I am getting the following error. Could you tell me if it is due to this newer image? The Jenkins job for nodebb deployment however has failed because of that, and the nodebb container is having CrashLoopBack continuously. Please find the logs below:

admin_@jenkins:~$ kubectl logs -f nodebb-7dc47dbcf6-49njx -n dev
2022-08-26T15:17:42.134Z [10] - info: NodeBB Setup Triggered via Command Line

Welcome to NodeBB v1.18.6!

This looks like a new installation, so you'll have to answer a few questions about your environment before we can proceed.
Press enter to accept the default setting (shown in brackets).
2022-08-26T15:17:42.459Z [10] - info:
Now configuring redis database:
2022-08-26T15:17:43.916Z [10] - verbose: [minifier] utilizing a maximum of 1 additional threads
2022-08-26T15:17:43.938Z [10] - warn: NodeBB Setup Aborted.
 ReplyError: MISCONF Redis is configured to save RDB snapshots, but it is currently not able to persist on disk. Commands that may modify the data set are disabled, because this instance is configured to report errors during writes if RDB snapshotting fails (stop-writes-on-bgsave-error option). Please check the Redis logs for details about the RDB error.
    at parseError (/usr/src/app/node_modules/redis-parser/lib/parser.js:179:12)
    at parseType (/usr/src/app/node_modules/redis-parser/lib/parser.js:302:14)

Starting NodeBB
  "./nodebb stop" to stop the NodeBB server
  "./nodebb log" to view server output
  "./nodebb help" for more commands

2022-08-26T15:17:45.237Z [4567/39] - info: Initializing NodeBB v1.18.6 http://0.0.0.0:4567/discussions/
2022-08-26T15:17:47.577Z [4567/39] - warn: [cache-buster] could not read cache buster ENOENT: no such file or directory, open '/usr/src/app/build/cache-buster' {"code":"ENOENT","errno":-2,"path":"/usr/src/app/build/cache-buster","stack":"Error: ENOENT: no such file or directory, open '/usr/src/app/build/cache-buster'","syscall":"open"}

Parsing upgrade scripts...
OK | 109 script(s) found
  → [2015/12/15] Upgrading chats... skipped
2022-08-26T15:17:47.673Z [4567/39] - error: ReplyError: MISCONF Redis is configured to save RDB snapshots, but it is currently not able to persist on disk. Commands that may modify the data set are disabled, because this instance is configured to report errors during writes if RDB snapshotting fails (stop-writes-on-bgsave-error option). Please check the Redis logs for details about the RDB error.
    at parseError (/usr/src/app/node_modules/redis-parser/lib/parser.js:179:12)
    at parseType (/usr/src/app/node_modules/redis-parser/lib/parser.js:302:14)

**Reply by vinukumar-vs** [reply]
@rajeevsathish  Can you please take a look at NodeBB build issues?
Let me know if you need any help on NodeBB & Plugins.

@reshmi-nair Take Ventak help if you there are any issues on NodeBB.

**Reply by beepdot** [reply]
@shounakacharya Please can you check the redis logs for more inforamtion, looks like redis is unable to save snapshots - `stop-writes-on-bgsave-error`

**Reply by shounakacharya** [reply]
@keshavprasadms : Yes. It was bg errors reported by Redis because it did not have permissions on the /data directory. Changing the permissions on the directory made it work. thanks. Marking this thread as answered!
